### PR TITLE
Kafka documentation refactor: Concepts -> Connection -> Messaging -> Details

### PR DESCRIPTION
**Pull request by arkadius**

### CODE REVIEW

File: docs/integration/KafkaIntegration.md [hunk 1]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
 
-## Schema Registry integration
+### Schemas
 
-Nussknacker integrates with Schema Registries. It is the source of knowledge about:
-- Which topics are available in Kafka sources and sinks
-- What schema versions are available for topic
-- How code completions and validations on types described by schemas should work
-- How messages will be serialized and deserialized
+Nussknacker uses schemas for code completions and validations of messages.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:27]: To me a bit too concise. What something like this 
-----
Schema defines the content and the format of data; Nussknacker expects that messages in the topics are described by the schema. Nussknacker uses information contained in schemas for code completion and validation of messages.

File: docs/integration/KafkaIntegration.md [hunk 2]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
-- How code completions and validations on types described by schemas should work
-- How messages will be serialized and deserialized
+Nussknacker uses schemas for code completions and validations of messages. 
+Schema of message can be described in [Avro Schema format](https://avro.apache.org/docs/#schemas) or [JSON Schema format](https://json-schema.org) (Confluent Schema Registry only)
 
-Currently, Nussknacker supports two implementations of schema registries: based on *Confluent Schema Registry* and based on *Azure Schema Registry*. 
-During runtime, it serializes and deserializes messages in the way that is compatible with standard Kafka serializers and deserializers 
-delivered by those schema registries providers. Thanks to that you should be able to send messages to Nussknacker and read messages 
-produced by Nussknacker using standard tooling available around those Schema Registries.
+Schemas are retrieved from Schema Registry - *Confluent Schema Registry* and *Azure Schema Registry* are supported.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:30]: are retrieved -> are stored 
                      -> are stored and retrieved

File: docs/integration/KafkaIntegration.md (hunk 1 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:27]: Thanks for remark, but why content AND format of data? How content is different from format of data? Isn't the same?

File: docs/integration/KafkaIntegration.md [hunk 3]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+
+#### Association between schema with topic
+
+To properly present information about topics and version and to recognize which schema is assigned to version, Nussknacker follow conventions:
+- For Confluent-based implementation it uses [TopicNameStrategy for subject names](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#subject-name-strategy).
+  It means that it looks for schemas available at `<topic-name>-(key or value)` *subject*. For example for topic `transactions`, it looks for schemas at `transactions-key` *subject* for key and `transactions-value` *subject* for value
+- In the Azure Schema Registry, *subject* concept doesn't exist - schemas are grouped by the same *schema name*. Because of that, Nussknacker introduces
+  own convention for association between schema and topic: *schema name* should be in format: `CamelCasedTopicNameKey` for keys and `CamelCasedTopicNameValue` for values.
+  For example for `input-events` topic, schema name should be named `InputEventsKey` for key or `InputEventsValue` for value. Be aware that it may require change of schema name
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:41]: _Be aware that ..._
If the intention is to say that _if before you  implemented Nu you already had had schema definitions and they do not conform to our convention, Nu will not be happy_ - I think it is obvious and can be omitted. If the intention was different, I did not understand it.

File: docs/integration/KafkaIntegration.md [hunk 4]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+  not only in Azure portal but also inside schema content - those names should be the same to make serialization works correctly
+
+## Connection and Authentication Configuration
 
-### Connection configuration
+Under the hood Nussknacker uses `kafkaProperties` to configure standard kafka client. It means that all standard Kafka client properties will be respected.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:46]: I would omit the second sentence, what follows is a clear indication of some _not to be ignored_ configuration. 
Alternatively, I would replace word _respected_ with _provided_.

File: docs/integration/KafkaIntegration.md [hunk 5]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+kafkaProperties {
+  "schema.registry.url": "http://schemaregistry:8081"
+  "bootstrap.servers": "broker1:9092,broker2:9092"
+  "security.protocol": "SASL_SSL"
+  "sasl.mechanism": "PLAIN"
+  "sasl.jaas.config": "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"some_user\" password=\"some_user\";"
+}
+```
+
+If you use Azure Events Hubs (which uses this mode), username will be `$ConnectionString` and password will be the connection string starting with `Endpoint=sb://`.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:69]: what is _this mode_ here? SASL_SSL?

File: docs/integration/KafkaIntegration.md [hunk 6]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+
+In case if you use your own CA and client+server certificates authentication, you should additionally provide:
+`ssl.keystore.location`, `ssl.keystore.password`, `ssl.key.password`, `ssl.truststore.location`, `ssl.truststore.password`.
+
+To make sure if your configuration is correct, you can test it with standard kafka-cli commands
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:74]: make sure if -> check if
or
make sure if -> make sure that

File: docs/integration/KafkaIntegration.md [hunk 7]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+- [Azure Event Hubs for Kafka quickstart](https://github.com/Azure/azure-event-hubs-for-kafka/tree/master/quickstart/kafka-cli)
+
+After you'll get properly working set of properties, you just need to copy it to Nussknacker's configuration.
+
+### Schema Registry - Connection
+
+To configure connection Schema Registry, you need to configure at least `schema.registry.url`. It should contain comma separated list of urls to Schema Registry.
+For the single node installation, it will be just an url. Be aware that contrary to Kafka brokers, Schema Registry urls should start with `https://` or `http://`.
+
+Currently, Nussknacker supports two implementations of Schema Registries: based on *Confluent Schema Registry* and based on *Azure Schema Registry*.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:88]: I would make this sentence a first sentence in this section

File: docs/integration/KafkaIntegration.md [hunk 8]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+Given implementation is picked based on `schema.registry.url` property. By default, Confluent-based implementation is used.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:89]: I would rephrase it to something like
_Nussknacker determines which registry implementation (Confluent or Azzure) is used from the schema.registry.url property. It the URL ends with .servicebus.windows.net Nussknacker assumes that Azure  schema registry is assumed; if not Confluent schema registry is assumed_.

File: docs/integration/KafkaIntegration.md [hunk 9]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
-- For Confluent-based implementation it uses [TopicNameStrategy for subject names](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#subject-name-strategy).
-  It means that it looks for schemas available at `\<topic-name\>-(key or value)`. For example for topic `transactions`, it looks for schemas at `transactions-key` for key and `transactions-value` subject for value
-- In Azure Schema Registry, subjects concept doesn't exist - schemas are grouped by the same schema name. Because of that, Nussknacker introduces
-  own convention for association between schema and topic: schema name should be in format: CamelCasedTopicNameKey for keys and CamelCasedTopicNameValue for values.
-  For example for `input-events` topic, schema name should be named `InputEventsKey` for key or `InputEventsValue` for value. Be aware that it may require change of schema name
-  not only in Azure portal but also inside schema content - those names should be the same to make serialization works correctly
-  
-### Schema and payload types
+To communicate with running Nussknacker's scenarios via Kafka sources and sinks you can use for example standard kafka-cli commands
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:114]: ---> You can use standard kafka-cli commands to interact (?? communicate sounds a bit odd to me) with kafka source and sink topics used in Nu scenarios. (I think more precise than sources and sinks)

File: docs/integration/KafkaIntegration.md [hunk 10]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+like `kafka-console-consumer`, `kafka-console-producer`, [kcat](https://github.com/edenhill/kcat), Confluent's
+`kafka-avro-console-consumer`, `kafka-avro-console-producer` commands for Confluent-based Avro encoded messages or graphical tools like [AKHQ](https://akhq.io).
+
+Be aware that Azure-based Avro encoded messages have a little different format than Confluent once - Schema ID is passed in headers instead of payload.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:118]: "once" is probably not needed

File: docs/integration/KafkaIntegration.md [hunk 11]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+produced by Nussknacker using standard tooling available around those Schema Registries. To see how those formats are different, take a look at [Schema Registry comparison section](#schema-registry-comparison)
+
+For some situations it might be helpful to use JSON payload with Avro schema. Especially when your Schema Registry doesn't support JSON schemas.
+You can do that by enabling `avroAsJsonSerialization` configuration setting.
 
 ### Schema ID
 
-Nussknacker supports schema evolution.
+Each topic can contain messages written using different schema versions. Schema versions are identified by *Schema ID*. 
+Nussknacker need to know what was the schema used during writing to make message validation and schema evolution possible.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:139]: need -> needs

File: docs/integration/KafkaIntegration.md [hunk 12]
```diff
@@ -8,104 +8,164 @@ sidebar_position: 1
...
+Because of that it need to extract *Schema ID* from the message.
```
Comment (ZbyszekMM) [inline docs/integration/KafkaIntegration.md:140]: it need -> Nussknacker needs

File: docs/integration/KafkaIntegration.md (hunk 1 above)
Comment (ZbyszekMM) [reply docs/integration/KafkaIntegration.md:27]: Could not find a reasonable - easy to understand (not too formal / scientific) definition of what schema is (in our context). Pls replace it with anything you deem reasonable and precise.

**Review by ZbyszekMM — APPROVED** [review]

File: docs/integration/KafkaIntegration.md (hunk 2 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:30]: I originally wrote sth like this, but after that I realize that it can be confusing, because someone can think that he/she can manage (add/remove) schemas in Nussknacker. But retrieved indeed looks "technical". Maybe Schemas "are managed" by Schema Registry?

File: docs/integration/KafkaIntegration.md (hunk 3 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:41]: I think that it is impossible to understand this notice without usage of Azure portal. There are schema names that are something different than schema name inside schema. I can show you if you want.

File: docs/integration/KafkaIntegration.md (hunk 4 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:46]: I want to write here, that we respect many options that are not documented in our code, but that are documented in other documentations. Is it clear without this sentence?

File: docs/integration/KafkaIntegration.md (hunk 5 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:69]: Yes, the mode just described before this sentence - SASL_SSL. Is it clear or should I change it somehow?

File: docs/integration/KafkaIntegration.md (hunk 6 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:74]: This step is not mandatory, because of that I've added "To". Should it be e.g "To make sure  that your ..." or should I change it in other way?

File: docs/integration/KafkaIntegration.md (hunk 7 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:88]: Ok, good point

File: docs/integration/KafkaIntegration.md (hunk 8 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:89]: ok

File: docs/integration/KafkaIntegration.md (hunk 9 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:114]: ok, changed

File: docs/integration/KafkaIntegration.md (hunk 11 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:139]: good catch

File: docs/integration/KafkaIntegration.md (hunk 12 above)
Comment (arkadius) [reply docs/integration/KafkaIntegration.md:140]: Nussknacker will be used twice but maybe it is not the problem, I'll change it.
